### PR TITLE
Improving perf of unwrapping object

### LINKED ISSUE #437: Performance of TryUnwrapObject can be improved by caching attributes.
When looking into XAML's Dependency Property performance, I noticed that there's a significant overhead coming from using reflection to get custom attributes in WinRT.ComWrappersSupport.TryUnwrapObject.  If these were cached, it could potentially reduce time spent in this method.  Stack info below:

image

### PR DESCRIPTION
**Pull request by manodasanW**
Improving perf of unwrapping objects by removing the use of reflection to get custom attributes to check for an ObjectReferenceWrapper and rather using IWinRTObject in NET 5 to get the same object.

- Everywhere we had an ObjectReferenceWrapper attribute and had implemented IWinRTObject, we now have a bool which indicates it has a native object to unwrap.
- The only exception to this is composable types which check whether the type has been composed / extended by a custom C# type and if not, it returns true and otherwise false which matches the previous behavior.
- For places which didn't have an ObjectReferenceWrapper attribute,  but implemented IWinRTObject,  it returns false.
- There are couple of custom projected ABI types which have the ObjectReferenceWrapper attribute, but don't implement IWinRTObject.  From inspection of them, they seem to project to C# system class types which means we are not probably unwrapping them back to the native object.  The ones which projected to C# system interface types were all dynamicinterfacecastable implementations other than for one which I converted.
- NativeDelegateWrapper is the other one I noticed which has an ObjectReferenceWrapper, but doesn't implement IWinRTObject.  I tried to implement a couple of tests which tried to cause them to unwrap to see whether I should implement  IWinRTObject on them (i.e. returning a delegate from a C++/WinRT component and then passing it back as a delegate and as an object).  In the first case, delegate marshaling doesn't seem to unwrap the native object at all and just creates a CCW on our managed delegate wrapper.  Is that intended or a bug?  And in the second case where I tried to pass the delegate as an object, it did unwrap before, but it then crashes as C++/WinRT delegates doesn't implement IInspectable apparently.  With my changes as present, it no longer unwraps the delegate, but creates a CCW on the wrapper which happens to succeed on the IInspectable QI and due to that doesn't crash,  This is unintentionally parity with delegate marshaling, but if  we expect it should be unwrapped, we should make NativeDelegateWrapper implement IWinRTObject and update delegate marshaling to also do unwrapping if any.  Thoughts?

Fixes #437

### CODE REVIEW

File: WinRT.Runtime/ComWrappersSupport.net5.cs [hunk 1]
```diff
@@ -80,8 +80,29 @@ public static T CreateRcwForComObject<T>(IntPtr ptr)
                 ABI.System.Nullable<Type> nt => (T)(object) nt.Value,
                 _ => (T) rcw
             };
-        }
-
+        }
+
+        public static bool TryUnwrapObject(object o, out IObjectReference objRef)
+        {
+            // The unwrapping here needs to be in exact type match in case the user
```
Comment (Scottj1s) [inline WinRT.Runtime/ComWrappersSupport.net5.cs:87]: "an exact"

File: WinRT.Runtime/ComWrappersSupport.netstandard2.0.cs [hunk 2]
```diff
@@ -75,8 +76,53 @@ public static T CreateRcwForComObject<T>(IntPtr ptr)
...
+                if (projectedClass is object && projectedClass.DefaultInterfaceProperty != null)
+                {
+                    return type.GetProperty(projectedClass.DefaultInterfaceProperty, BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.DeclaredOnly);
+                }
+
+                return null;
+            });
+
+
+            if (objRefField is FieldInfo field)
```
Comment (Scottj1s) [inline WinRT.Runtime/ComWrappersSupport.netstandard2.0.cs:112]: you could avoid the type checks by keeping a cache of System.Funcs (factories) that take in o and return objRef

**Review by Scottj1s — CHANGES_REQUESTED** [review]
change to TypeObjectRefAccesorCache = new ConcurrentDictionary<Type, Func<object, IObjectReference>>

**Comment by Scottj1s** [comment]
> * should we make NativeDelegateWrapper implement IWinRTObject and update delegate marshaling to also do unwrapping

my naive answer is "yes" - any downsides?

File: WinRT.Runtime/ComWrappersSupport.netstandard2.0.cs (hunk 2 above)
Comment (Scottj1s) [reply WinRT.Runtime/ComWrappersSupport.netstandard2.0.cs:112]: I guess it was this way to start with - but would still help perf a bit

Comment (manodasanW) [reply WinRT.Runtime/ComWrappersSupport.netstandard2.0.cs:112]: Yea I already have the changes to address this feedback.  Working on fixing delegate marshaling.

**Review by Scottj1s — APPROVED** [review]

**Pull request merged**